build-tests/heft-example-plugin-02/src/index.ts

```diff
@@ -0,0 +1,20 @@
+import { IHeftPlugin, HeftSession, HeftConfiguration } from '@rushstack/heft';
+import { default as heftExamplePlugin01 } from 'heft-example-plugin-01';
+
+const PLUGIN_NAME: string = 'example-plugin-02';
+
+export class ExamplePlugin02 implements IHeftPlugin {
+  public displayName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    heftSession.applyForPlugin(heftExamplePlugin01, (plugin: typeof heftExamplePlugin01) => {
+      plugin.hooks.exampleHook.tap(PLUGIN_NAME, () => {
+        heftConfiguration.globalTerminal.writeLine(
+          `!!!!!!!!!!!!!!! Plugin "${plugin.displayName}" hook called !!!!!!!!!!!!!!! `
```

Comment:
@iclanton Here is a sketch of the above ideas:

**heft-example-plugin-01/src/index.ts**
```ts
export interface IExamplePlugin01Hooks {
  exampleHook: SyncHook;
}

/**
 * This defines the public contract for other plugins that want to access ExamplePlugin01.
 * Any changes to changes to this interface could break consumers.
 */
export interface IExamplePlugin01Accessor extends IHeftPluginAccessor {
  pluginId: string;
  hooks: IExamplePlugin01Hooks;
}

// This allows Heft to discover plugins without executing all their code.
export const examplePlugin01Manifest: IHeftPluginManifest = {
  pluginId: 'example-plugin-01',
  displayName: 'Example 1',
  entryPoint: './ExamplePlugin01.js'
};
```

**heft-example-plugin-01/src/ExamplePlugin01.ts**
```ts
// This object is private implementation details. Other plugins are NOT allowed to access it.
export class ExamplePlugin01 implements IHeftPlugin {
  public readonly displayName: string = examplePlugin01Manifest.displayName;

  public readonly hooks: IExamplePlugin01Hooks;

  constructor() {
    this.hooks = {
      exampleHook: new SyncHook()
    };
  }

  // Heft will call this at most once, if/when another plugin uses requestAccessToPlugin()
  // to access this plugin.  If `createAccessor` is not defined, then an error will be reported
  // saying that ExamplePlugin01 does not agree to be extended.
  public createAccessor(): IExamplePlugin01Accessor {
    return {
      manifest: examplePlugin01Manifest,
      hooks: this.hooks
    };
  }

  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
    heftSession.hooks.build.tap(examplePlugin01Manifest.pluginId, (build: IBuildStageContext) => {
      build.hooks.preCompile.tap(examplePlugin01Manifest.pluginId, (preCompile: IPreCompileSubstage) => {
        preCompile.hooks.run.tap(examplePlugin01Manifest.pluginId, () => {
          this.hooks.exampleHook.call();
        });
      });
    });
  }
}
```

**heft-example-plugin-02/src/index.ts**
```ts
// Note that we do NOT import the ExamplePlugin01 class.
// We only import the manifest metadata and the "public API"
import { examplePlugin01Manifest, IExamplePlugin01Accessor } from 'heft-example-plugin-01';

// This allows Heft to discover plugins without executing all their code.
export const examplePlugin02Manifest: IHeftPluginManifest = {
  pluginId: 'example-plugin-02',
  displayName: 'Example 2',
  entryPoint: './ExamplePlugin02.js'
};


export class ExamplePlugin02 implements IHeftPlugin {
  public displayName: string = examplePlugin02Manifest.pluginName;

  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
    // Here we ask Heft to locate the instance of "heft-example-plugin-01" and return its accessor.
    // If the plugin is not configured to be loaded, then Heft could report an error.
    heftSession.requestAccessToPlugin(examplePlugin01Manifest.pluginId, (examplePlugin01: IExamplePlugin01Accessor) => {
      examplePlugin01.hooks.exampleHook.tap(examplePlugin02Manifest.pluginId, () => {
        heftConfiguration.globalTerminal.writeLine(
          `!!!!!!!!!!!!!!! Plugin "${examplePlugin01.pluginId}" hook called !!!!!!!!!!!!!!! `
        );
      });
    });
  }
}
```

Reply:
With the caveat that I still feel that calling `tap()` or `requestAccessToPlugin()` is a very important activity that Heft needs to oversee.  The caller should not need to constantly explain who they are.

For example, this code:
```ts
    heftSession.requestAccessToPlugin(
      PluginIds.ExamplePlugin02,  // who is attaching
      OtherPluginIds.ExamplePlugin01,  // whom we are attaching to
      (examplePlugin01: IExamplePlugin01Accessor) => {
      examplePlugin01.hooks.exampleHook.tap(PluginIds.ExamplePlugin02, () => {
```
...could be reduced to an API where the `this`=`PluginIds.ExamplePlugin02` so we don't need to specify our own `pluginId`:

```ts
    this.requestAccessToPlugin(OtherPluginIds.ExamplePlugin01,
      (examplePlugin01: IExamplePlugin01Accessor) => {
      this.attachHandler(examplePlugin01.hooks.exampleHook, () => {
```

Reply:
It feels very awkward to enforce that plugins are classes extending from a base class because that means that plugins are married to a particular version of Heft and that plugins need to take a dependency on Heft. I'd rather keep the contract simple and let plugins be isolated with the theory that most plugins will perform simple/self-contained operations.